llvm/lib/Transforms/IPO/ArgumentPromotion.cpp

```diff
@@ -485,11 +485,39 @@ static bool allCallersPassValidPointerForArgument(
   });
 }
 
+// Try to prove that all Calls to F do not modify the memory pointed to by Arg,
+// using alias analysis local to each caller of F.
+static bool isArgUnmodifiedByAllCalls(Function *F, unsigned ArgNo,
+                                      FunctionAnalysisManager &FAM) {
+  // Check if all Users of F are Calls which do not modify Arg.
+  for (User *U : F->users()) {
+
+    // Bail if we find an unexpected (non CallInst) use of the function.
+    auto *Call = dyn_cast<CallInst>(U);
+    if (!Call)
+      return false;
+
+    Value *ArgOp = Call->getArgOperand(ArgNo);
+    assert(ArgOp->getType()->isPointerTy() && "Argument must be Pointer Type!");
```

Comment:
I don't think its possible for Arg to not be a pointer type at this stage, since `PointerArgs` in `promoteArguments`. I think this check can be deleted.

Reply:
Fair point - I've removed this assert in [Address review comments 2](https://github.com/llvm/llvm-project/pull/106216/commits/aa809e9dc8a9bf5b5fc0cdc6f56549bd7806879d)